### DISCUSSION TITLE
7.0.0-preview5

### DISCUSSION THREAD

**Discussion by jennyf19** [AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet#2294] [IdentityModel 7x and .NET 8]
7.0.0-preview5
=======
### Bug fixes:
- Improve log messages. See PR [#2289](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2289) for details.
- In `AadIssuerValidator` return a `ValueTask<string>` instead of a `Task<string>`. See Issue [#2286](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2286) and PR [https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2287] for details.
- Deprecate `int? JwtPayload.Exp`, `.Iat`, and `.Nbf`. See issue [#2266](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2266) for details.
- General clean-up. See PR [#2285](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2285).

<hr /><em>This discussion was created from the release <a href='https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/releases/tag/7.0.0-preview5'>7.0.0-preview5</a>.</em>

**Comment by jennyf19** [comment]
@kevinchalet @leastprivilege @brockallen @cakescience 
 
 We are hoping this will be our last preview before we GA 7.0.0 later this week. Thanks all for your testing, feedback, and quick opening of issues in previous versions.

**Reply by leastprivilege** [reply]
We are seeing dozens of failing tests in IdentityServer when switching to v7. Give us a couple of days to investigate. thanks.

**Reply by josephdecock** [reply]
The main issue I'm seeing in IdentityServer's tests are `TypeLoadException`s that seem to be a dependency conflict with Microsoft.AspNetCore.Authentication.OpenIdConnect.

To quickly reproduce the problem outside IdentityServer:
- create a new ASP.NET Core Web App
- add Microsoft.AspNetCore.Authentication.OpenIdConnect and configure authentication (I used demo.duendesoftware.com as my authority):
```
builder.Services.AddAuthentication(opt =>
    {
        opt.DefaultScheme = "Cookie";
        opt.DefaultChallengeScheme = "demo";
    })
    .AddCookie("Cookie")
    .AddOpenIdConnect("demo", options =>
    {
        options.SignInScheme = "Cookie";
        options.Authority = "https://demo.duendesoftware.com";
        options.ClientId = "login";
        options.ResponseType = "id_token";
    });
```
- Add a reference to System.IdentityModel.Tokens.Jwt
```
<PackageReference Include="System.IdentityModel.Tokens.Jwt" Version="7.0.0-preview5" />
```
- Finally, issue a challenge (I added `[Authorize]` to the privacy page). 

This results in the following exception:

```
TypeLoadException: Could not load type 'Microsoft.IdentityModel.Json.JsonConvert' from assembly 'Microsoft.IdentityModel.Tokens, Version=7.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.

    Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(string address, IDocumentRetriever retriever, CancellationToken cancel)
    System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start<TStateMachine>(ref TStateMachine stateMachine)
    System.Runtime.CompilerServices.AsyncTaskMethodBuilder<TResult>.Start<TStateMachine>(ref TStateMachine stateMachine)
    Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.GetAsync(string address, IDocumentRetriever retriever, CancellationToken cancel)
    Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfigurationRetriever.Microsoft.IdentityModel.Protocols.IConfigurationRetriever<Microsoft.IdentityModel.Protocols.OpenIdConnect.OpenIdConnectConfiguration>.GetConfigurationAsync(string address, IDocumentRetriever retriever, CancellationToken cancel)
    Microsoft.IdentityModel.Protocols.ConfigurationManager<T>.GetConfigurationAsync(CancellationToken cancel)
```

**Reply by josephdecock** [reply]
Oh, and for additional context, my `dotnet --version` is 8.0.100-preview.7.23376.3.

**Reply by josephdecock** [reply]
Thanks! That clears up the dependency issues I was seeing. We still have some other failing tests that I'm working through in IdentityServer.

**Reply by jennyf19** [reply]
@josephdecock any update on your end? Were you able to get the failing tests resolved?

**Reply by josephdecock** [reply]
Things are looking good in IdentityServer. 

The most significant thing that I ran into was that `TryGetHeaderValue<T>` now [no longer supports `IDictionary<string, object>` as the type parameter](https://github.com/DuendeSoftware/IdentityServer/pull/1401/files#diff-de802ee67e287ae21081141470e8708a90da8e196dd7be4d6dd9bac8b9e1e017L157). I went through a few different iterations of this to understand the types that are supported and make a good choice here, and I would love to see documentation about it. Questions that I think users will have include
- Should I use `Dictionary<string,object>`, or `JsonElement`,  or something else? And,
- When I do have a Dictionary<string,object>, what types can the object be at run time?

The other issues we saw were largely to do with test setup:
- Invalid json with single quotes instead of double quotes is now rejected by the `JsonWebKey` constructor. This seems totally reasonable. Single quotes obviously aren't valid json, it was just convenient for creating test objects, but there are [other ways](https://github.com/DuendeSoftware/IdentityServer/pull/1401/files#diff-3af36f88ff4eb78841a4e87cb9907a0215b24c15a094d5ec668cb173c0eaea5cL39) of doing the same thing that I think actually improved legibility.
- Passing anonymous objects to `CreateToken` [within the header dictionary](https://github.com/DuendeSoftware/IdentityServer/pull/1401/files#diff-23e540be09f8febfc661a733fe2e4d3c3de98482dea3de9019ea41c7d0789babL173-L179) is not supported. Again, that wasn't something we wanted to do in real code, it was just a quick way to create the headers in a test, so if this is a compromise that is needed, that's fine.
